src/hotspot/share/opto/addnode.cpp

```diff
@@ -304,14 +304,14 @@ Node* AddNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
     }
   }
 
-  // Convert "x+(0-y)" into "(x-y)"
-  if (op2 == Op_Sub(bt) && phase->type(in2->in(1)) == TypeInteger::zero(bt)) {
-    return SubNode::make(in1, in2->in(2), bt);
+  // Convert (con - y) + x into "(x - y) + con"
+  if (op1 == Op_Sub(bt) && in1->in(1)->Opcode() == Op_ConIL(bt)) {
+    return AddNode::make(phase->transform(SubNode::make(in2, in1->in(2), bt)), in1->in(1), bt);
   }
 
-  // Convert "(0-y)+x" into "(x-y)"
-  if (op1 == Op_Sub(bt) && phase->type(in1->in(1)) == TypeInteger::zero(bt)) {
-    return SubNode::make(in2, in1->in(2), bt);
+  // Convert x + (con - y) into "(x - y) + con"
+  if (op2 == Op_Sub(bt) && in2->in(1)->Opcode() == Op_ConIL(bt)) {
+    return AddNode::make(phase->transform(SubNode::make(in1, in2->in(2), bt)), in2->in(1), bt);
```

Comment:
Both of proposed transformations are similar one in `AddNode::Ideal()` lines 181-214.
There is exception there to avoid transformation if (y+con) could be increment for loop's variable. I wonder if we need the same check here.

Reply:
Thanks for reviewing! That is indeed an interesting observation. I feel `con - y` could not be a loop increment so it should be sufficient to check `x` is not a loop induction variable?

Reply:
I think we need to check `y`. For example:

```
for (int i = 0; i < 10000; i++) {
    a[MAX - i] = b[MAX - i];
}
```

Reply:
But in this case it seems OK to me to make the transformation, for example:
```
for (int i = 0; i < 10000; i++) {
    a[x + (MAX - i)] = b[x + (MAX - i)]; // a[(x - i) + MAX] = b[(x - i) + MAX]
}
```
Is there something I overlooked?

Reply:
In loopopts (and superword) we are loking for `(+-phi  +- const)` pattern for loop variable.
With your transformation we get `((+-phi +- x) + const)` which may prevent some optimizations.

Reply:
Makes sense. Thanks for the explanation. I included the check in a new commit.